### PR TITLE
Put limits on rqt data upload

### PR DESCRIPTION
**Pull request by lfitchett**
This will not upload rqt data older than 1 week (default value, configurable with MetricMaxUploadAge).

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs [hunk 1]
```diff
@@ -27,23 +31,54 @@ public Task StoreMetricsAsync(IEnumerable<Metric> metrics)
...
+        async IAsyncEnumerable<Metric> GetAllMetricsInternal([EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            HashSet<long> maxAgeEntriesToRemove = new HashSet<long>();
+
+            foreach ((long offset, IEnumerable<Metric> metrics) in await this.dataStore.GetBatch(0, 1000, cancellationToken))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                // This makes the assumption that metrics are stored in batches with matching timestamps.
+                // This is the case b/c the metrics are stored immediately after being scraped.
+                if (metrics.First().TimeGeneratedUtc < DateTime.UtcNow - this.maxMetricAge)
```
Comment (damonbarry) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs:47]: `UtcNow` will be reevaluated each time through the loop, so different batches will be evaluated against different max-age timestamps. Is that the right behavior or do we want to take a snapshot of `UtcNow` before we enter the loop?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs [hunk 2]
```diff
@@ -27,23 +31,54 @@ public Task StoreMetricsAsync(IEnumerable<Metric> metrics)
...
+                {
+                    // TODO: log removed
```
Comment (damonbarry) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs:49]: Can we add the logging in this PR? I imagine we'd count up the total # of removed entries and report it in a single log statement (at least # removed entries, although time range would be nice to have)

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs [hunk 3]
```diff
@@ -27,23 +31,54 @@ public Task StoreMetricsAsync(IEnumerable<Metric> metrics)
...
+                }
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // remove metrics past ttl
+            await this.RemoveOffsets(maxAgeEntriesToRemove);
+        }
+
+        public Task RemoveAllReturnedMetricsAsync()
```
Comment (damonbarry) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs:68]: Remove the `Async` suffix?

Comment (lfitchett) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs:68]: Its still a task, I thought we named it async if it returns a task. Thats all the caller cares about anyway.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs (hunk 1 above)
Comment (lfitchett) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs:47]: It should happen so fast it won't matter (we're comparing on a week timescale), but I'll go ahead and change it for consistency

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs (hunk 2 above)
Comment (lfitchett) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs:49]: added logging. I didn't bother logging at the end. The only time multiple scrapes will be deleted at once is if the device was powered off for an extended period of time.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs (hunk 3 above)
Comment (damonbarry) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs:72]: You're right. Not sure what I was thinking...

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs (hunk 2 above)
Comment (damonbarry) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs]: Looks good.

**Review by damonbarry — APPROVED** [review]

**Review by damonbarry — APPROVED** [review]

**Pull request merged**